### ISSUE TITLE
bug in normal.gee?

### ISSUE THREAD

**Issue by andreashandel** [IQSS/Zelig#244]
I'm trying to fit a normal.gee model. If I do it directly with the geeglm function, it works. If I do what I think is exactly the same thing through Zelig, it fails. Here is a reproducible example:

```R
mydata=mtcars; #use mtcars data

mydata$myid = sample(1:10,size=nrow(mydata),replace=TRUE) #make some fake cluster ID
mydata <- mydata %>% dplyr::arrange( myid) #sort by cluster ID

geepack::geeglm(formula = mpg ~ ., family = gaussian, data = mydata, id = mydata$myid) #this works

Zelig::zelig(formula = mpg ~ ., model = "normal.gee", id = "myid", data = mydata) #this fails

## Error is: Error in geepack::geeglm(formula = formula, id = c(1L, 2L, 2L, 2L, 2L,  : 
##   Model matrix is rank deficient; geeglm can not proceed
```
Am I missing something obvious? Thanks.

**Comment by christophergandrud** [comment]
Thanks for your message @andreashandel. I'll take a look.

**Comment by christophergandrud** [comment]
Initial thought: it seems like the issue is with the way Zelig is handling the formula. For example, if you add in the right hand side variables explicitly it seems to work.  

I'll do more digging thought.

**Comment by christophergandrud** [comment]
I have a working solution in this branch: https://github.com/IQSS/Zelig/tree/dotformula

Download it with:

```r
devtools::install_github("IQSS/Zelig", ref = "dotformula")
```

I haven't merged it into master yet, however, because the solution seems to interfere with `relogit`. Will keep working on this.

**Comment by christophergandrud** [comment]
I have a working solution in this branch: https://github.com/IQSS/Zelig/tree/dotformula

Download it with:

```r
devtools::install_github("IQSS/Zelig", ref = "dotformula")
```

I haven't merged it into master yet, however, because the solution seems to interfere with `relogit`. Will keep working on this.

**Comment by christophergandrud** [comment]
Issue related to: https://github.com/IQSS/Zelig/issues/244

**Comment by christophergandrud** [comment]
Issue related to: https://github.com/IQSS/Zelig/issues/87

**Comment by andreashandel** [comment]
Thanks! 

I have a more general question (which changes the topic a bit, so I could open another issue with this if you think that's better.)

I want to fit/predict with normal.gee in a cross-validation framework. To that end, I want to fit my training data with normal.gee and then predict the test data outcomes. 
Something like this:
geefit = zelig(formula = outcome ~ . , model = "normal.gee", id = "myid", data = traindata)
newpredictors = setx(geefit, data = testdata)
geepred = sim(obj = geefit, x = newpredictors ) 
  
And then I'd like to pull out predictions for the outcomes in my test data set from geepred. So far I haven't been able to pull what I need out of what sim() returns, despite reading through the documentation. I'm wondering if I'm missing something here and maybe what I"m trying to do just won't work? (I'm using Zelig since both gee and geepack don't do predictions, and I was hoping sim() could give me what I wanted.) Thanks.

**Comment by christophergandrud** [comment]
We recently introduced a new function--`zelig_qi_to_df`--to extract simulations from `sim` in a data frame. (We are rewriting the docs, so this hasn't been fully included yet in the docs). 

Could you give this a test and if it doesn't work, open a separate issue?

**Comment by andreashandel** [comment]
Thanks for the quick reply! I tried the function, it seems to work, but I'm still not able to get what I want. I think I'm not fully understanding how things work. I assumed that in the example I gave above, if traindata is an KxN dataframe and testdata is an MxN dataframe, when I do setx() and then sim() as shown above, I get predictions for all M observations in my testdata. But looking at the output from sim(), there are 1000 predictions for a single set of predictor variables. Does that mean setx() and sim() are not vectorized and I would need to write a loop over each new observation in my testdata that I want to predict, then run setx() and predict() on each of them, i.e. M times? Or am I still getting it all wrong? Unfortunately I don't seem to be able to find the information I need to figure this out in the docs. (Also, in the end I only want a single prediction for my outcomes, so I would probably take the mean over the simulations returned by sim() )? Thanks!

**Comment by christophergandrud** [comment]
`setx` will make draw, by default,  1000 draws for each scenario given it. The scenarios are set by providing values for each covariate (and central tendencies for covariates not specified explicitly in `setx`), e.g.:

```r
setx(z.out, dist = 1:10)
```

This will make 1000 draws for each of 10 scenarios. 

I'm a little surprised that providing a data frame of fitted values didn't produce an error. I'll look into that. But otherwise you should try `setx` by proving values for each variable.

**Comment by christophergandrud** [comment]
Merged into master in b6a7a3a6124a57eb4a121ddc10d04d85153826ce
